File: rest/cursor_api/database/operations.py
"""Database write operations for chat and messages"""

import json
import sqlite3
from datetime import datetime
from typing import Dict, Any, List, Tuple


def ensure_chat_exists(
    conn: sqlite3.Connection,
    chat_id: str
) -> Tuple[bool, Dict[str, Any]]:
    """Ensure chat metadata exists in database, create if missing
    
    This implements the auto-create pattern from SKILL.mdc to handle the case
    where cursor-agent create-chat returns a UUID but doesn't create the database entry.
    
    Args:
        conn: SQLite connection
        chat_id: Chat UUID
        
    Returns:
        Tuple of (was_created: bool, metadata: dict)
        - was_created: True if chat was created, False if it already existed
        - metadata: The chat metadata dictionary
    """
    cursor = conn.cursor()
    
    # Check if chat exists
    cursor.execute(
        "SELECT value FROM cursorDiskKV WHERE key = ?",
        (f'composerData:{chat_id}',)
    )
    existing_chat = cursor.fetchone()
    
    if existing_chat:
        # Chat exists, return existing metadata
        metadata = json.loads(existing_chat[0])
        return False, metadata
    
    # Chat doesn't exist - create minimal metadata
    # This happens when using /agent/create-chat which only generates an ID
    now_ms = int(datetime.now().timestamp() * 1000)
    
    chat_metadata = {
        "_v": 10,
        "composerId": chat_id,
        "name": "Untitled",
        "richText": json.dumps({
            "root": {
                "children": [{
                    "children": [],
                    "format": "",
                    "indent": 0,
                    "type": "paragraph",
                    "version": 1
                }],
                "format": "",
                "indent": 0,
                "type": "root",
                "version": 1
            }
        }),
        "hasLoaded": True,
        "text": "",
        "fullConversationHeadersOnly": [],
        "createdAt": now_ms,
        "lastUpdatedAt": now_ms,
        "isArchived": False,
        "isDraft": False,
        "totalLinesAdded": 0,
        "totalLinesRemoved": 0,
    }
    
    cursor.execute(
        "INSERT INTO cursorDiskKV (key, value) VALUES (?, ?)",
        (f'composerData:{chat_id}', json.dumps(chat_metadata))
    )
    conn.commit()
    
    return True, chat_metadata


def save_message_to_db(
    conn: sqlite3.Connection,
    chat_id: str,
    bubble_id: str,
    bubble_data: Dict[str, Any]
) -> None:
    """Save a message bubble to the database
    
    Args:
        conn: SQLite connection
        chat_id: Chat UUID
        bubble_id: Message bubble UUID
        bubble_data: Complete bubble data structure
    """
    cursor = conn.cursor()
    key = f'bubbleId:{chat_id}:{bubble_id}'
    value = json.dumps(bubble_data)
    
    cursor.execute(
        "INSERT INTO cursorDiskKV (key, value) VALUES (?, ?)",
        (key, value)
    )


def update_chat_metadata(
    conn: sqlite3.Connection,
    chat_id: str,
    new_bubble_ids: List[tuple]  # [(bubble_id, type), ...]
) -> None:
    """Update chat metadata to include new messages
    
    Ensures all required fields are present for Cursor IDE compatibility.
    
    Args:
        conn: SQLite connection
        chat_id: Chat UUID
        new_bubble_ids: List of tuples (bubble_id, message_type)
        
    Raises:
        ValueError: If chat not found
    """
    cursor = conn.cursor()
    
    # Get existing metadata
    cursor.execute(
        "SELECT value FROM cursorDiskKV WHERE key = ?",
        (f'composerData:{chat_id}',)
    )
    row = cursor.fetchone()
    if not row:
        raise ValueError(f"Chat {chat_id} not found")
    
    metadata = json.loads(row[0])
    
    # Ensure critical fields exist
    if '_v' not in metadata:
        metadata['_v'] = 10
    
    if 'hasLoaded' not in metadata:
        metadata['hasLoaded'] = True
    
    if 'text' not in metadata:
        metadata['text'] = ""
    
    # Ensure richText exists (Lexical editor state)
    if 'richText' not in metadata:
        metadata['richText'] = json.dumps({
            "root": {
                "children": [{
                    "children": [],
                    "format": "",
                    "indent": 0,
                    "type": "paragraph",
                    "version": 1
                }],
                "format": "",
                "indent": 0,
                "type": "root",
                "version": 1
            }
        })
    
    # Add new messages to headers
    headers = metadata.get('fullConversationHeadersOnly', [])
    for bubble_id, msg_type in new_bubble_ids:
        headers.append({
            "bubbleId": bubble_id,
            "type": msg_type
        })
    metadata['fullConversationHeadersOnly'] = headers
    
    # Update timestamp
    metadata['lastUpdatedAt'] = int(datetime.now().timestamp() * 1000)
    
    # Save back
    cursor.execute(
        "UPDATE cursorDiskKV SET value = ? WHERE key = ?",
        (json.dumps(metadata), f'composerData:{chat_id}')
    )

File: rest-rust/src/db/operations.rs
use crate::error::Result;
use rusqlite::{Connection, Transaction};
use serde_json::{json, Value};
use std::collections::HashMap;

/// Ensure a chat exists in the database, creating it if necessary
pub fn ensure_chat_exists(
    conn: &Connection,
    chat_id: &str,
) -> Result<(bool, HashMap<String, Value>)> {
    let key = format!("composerData:{}", chat_id);
    let mut stmt = conn.prepare("SELECT value FROM cursorDiskKV WHERE key = ?")?;
    
    let existing = stmt.query_row([&key], |row| {
        let value_str: String = row.get(0)?;
        Ok(value_str)
    });
    
    match existing {
        Ok(value_str) => {
            let metadata: HashMap<String, Value> = serde_json::from_str(&value_str)?;
            Ok((false, metadata))
        }
        Err(rusqlite::Error::QueryReturnedNoRows) => {
            // Create minimal metadata
            let now_ms = chrono::Utc::now().timestamp_millis();
            
            // Create rich text structure
            let rich_text = json!({
                "root": {
                    "children": [{
                        "children": [],
                        "format": "",
                        "indent": 0,
                        "type": "paragraph",
                        "version": 1
                    }],
                    "format": "",
                    "indent": 0,
                    "type": "root",
                    "version": 1
                }
            });
            
            let mut metadata = HashMap::new();
            metadata.insert("_v".to_string(), Value::Number(10.into()));
            metadata.insert("composerId".to_string(), Value::String(chat_id.to_string()));
            metadata.insert("name".to_string(), Value::String("Untitled".to_string()));
            metadata.insert("richText".to_string(), Value::String(rich_text.to_string()));
            metadata.insert("hasLoaded".to_string(), Value::Bool(true));
            metadata.insert("text".to_string(), Value::String(String::new()));
            metadata.insert("fullConversationHeadersOnly".to_string(), Value::Array(vec![]));
            metadata.insert("createdAt".to_string(), Value::Number(now_ms.into()));
            metadata.insert("lastUpdatedAt".to_string(), Value::Number(now_ms.into()));
            metadata.insert("isArchived".to_string(), Value::Bool(false));
            metadata.insert("isDraft".to_string(), Value::Bool(false));
            metadata.insert("totalLinesAdded".to_string(), Value::Number(0.into()));
            metadata.insert("totalLinesRemoved".to_string(), Value::Number(0.into()));
            
            // Save to database
            let metadata_str = serde_json::to_string(&metadata)?;
            conn.execute(
                "INSERT INTO cursorDiskKV (key, value) VALUES (?, ?)",
                [&key, &metadata_str],
            )?;
            
            Ok((true, metadata))
        }
        Err(e) => Err(e.into()),
    }
}

/// Ensure a chat exists within a transaction
pub fn ensure_chat_exists_tx(
    tx: &Transaction,
    chat_id: &str,
) -> Result<(bool, HashMap<String, Value>)> {
    let key = format!("composerData:{}", chat_id);
    let mut stmt = tx.prepare("SELECT value FROM cursorDiskKV WHERE key = ?")?;
    
    let existing = stmt.query_row([&key], |row| {
        let value_str: String = row.get(0)?;
        Ok(value_str)
    });
    
    match existing {
        Ok(value_str) => {
            let metadata: HashMap<String, Value> = serde_json::from_str(&value_str)?;
            Ok((false, metadata))
        }
        Err(rusqlite::Error::QueryReturnedNoRows) => {
            // Create minimal metadata (same as above)
            let now_ms = chrono::Utc::now().timestamp_millis();
            
            let rich_text = json!({
                "root": {
                    "children": [{
                        "children": [],
                        "format": "",
                        "indent": 0,
                        "type": "paragraph",
                        "version": 1
                    }],
                    "format": "",
                    "indent": 0,
                    "type": "root",
                    "version": 1
                }
            });
            
            let mut metadata = HashMap::new();
            metadata.insert("_v".to_string(), Value::Number(10.into()));
            metadata.insert("composerId".to_string(), Value::String(chat_id.to_string()));
            metadata.insert("name".to_string(), Value::String("Untitled".to_string()));
            metadata.insert("richText".to_string(), Value::String(rich_text.to_string()));
            metadata.insert("hasLoaded".to_string(), Value::Bool(true));
            metadata.insert("text".to_string(), Value::String(String::new()));
            metadata.insert("fullConversationHeadersOnly".to_string(), Value::Array(vec![]));
            metadata.insert("createdAt".to_string(), Value::Number(now_ms.into()));
            metadata.insert("lastUpdatedAt".to_string(), Value::Number(now_ms.into()));
            metadata.insert("isArchived".to_string(), Value::Bool(false));
            metadata.insert("isDraft".to_string(), Value::Bool(false));
            metadata.insert("totalLinesAdded".to_string(), Value::Number(0.into()));
            metadata.insert("totalLinesRemoved".to_string(), Value::Number(0.into()));
            
            let metadata_str = serde_json::to_string(&metadata)?;
            tx.execute(
                "INSERT INTO cursorDiskKV (key, value) VALUES (?, ?)",
                [&key, &metadata_str],
            )?;
            
            Ok((true, metadata))
        }
        Err(e) => Err(e.into()),
    }
}

/// Save a message bubble to the database
pub fn save_message_to_db(
    conn: &Connection,
    chat_id: &str,
    bubble_id: &str,
    bubble_data: &HashMap<String, Value>,
) -> Result<()> {
    let key = format!("bubbleId:{}:{}", chat_id, bubble_id);
    let value_str = serde_json::to_string(bubble_data)?;
    
    conn.execute(
        "INSERT OR REPLACE INTO cursorDiskKV (key, value) VALUES (?, ?)",
        [&key, &value_str],
    )?;
    
    Ok(())
}

/// Save a message bubble within a transaction
pub fn save_message_to_db_tx(
    tx: &Transaction,
    chat_id: &str,
    bubble_id: &str,
    bubble_data: &HashMap<String, Value>,
) -> Result<()> {
    let key = format!("bubbleId:{}:{}", chat_id, bubble_id);
    let value_str = serde_json::to_string(bubble_data)?;
    
    tx.execute(
        "INSERT OR REPLACE INTO cursorDiskKV (key, value) VALUES (?, ?)",
        [&key, &value_str],
    )?;
    
    Ok(())
}

/// Update chat metadata
pub fn update_chat_metadata(
    conn: &Connection,
    chat_id: &str,
    updates: HashMap<String, Value>,
) -> Result<()> {
    let key = format!("composerData:{}", chat_id);
    
    // Get existing metadata
    let mut metadata: HashMap<String, Value> = {
        let mut stmt = conn.prepare("SELECT value FROM cursorDiskKV WHERE key = ?")?;
        let value_str: String = stmt.query_row([&key], |row| row.get(0))?;
        serde_json::from_str(&value_str)?
    };
    
    // Apply updates
    for (k, v) in updates {
        metadata.insert(k, v);
    }
    
    // Save back
    let metadata_str = serde_json::to_string(&metadata)?;
    conn.execute(
        "INSERT OR REPLACE INTO cursorDiskKV (key, value) VALUES (?, ?)",
        [&key, &metadata_str],
    )?;
    
    Ok(())
}

/// Update chat metadata within a transaction
pub fn update_chat_metadata_tx(
    tx: &Transaction,
    chat_id: &str,
    updates: HashMap<String, Value>,
) -> Result<()> {
    let key = format!("composerData:{}", chat_id);
    
    // Get existing metadata
    let mut metadata: HashMap<String, Value> = {
        let mut stmt = tx.prepare("SELECT value FROM cursorDiskKV WHERE key = ?")?;
        let value_str: String = stmt.query_row([&key], |row| row.get(0))?;
        serde_json::from_str(&value_str)?
    };
    
    // Apply updates
    for (k, v) in updates {
        metadata.insert(k, v);
    }
    
    // Save back
    let metadata_str = serde_json::to_string(&metadata)?;
    tx.execute(
        "INSERT OR REPLACE INTO cursorDiskKV (key, value) VALUES (?, ?)",
        [&key, &metadata_str],
    )?;
    
    Ok(())
}

/// Perform a complete message persistence operation with transaction
/// This ensures atomic writes: user message + AI response
pub fn persist_conversation_turn(
    conn: &mut Connection,
    chat_id: &str,
    user_bubble_id: &str,
    user_bubble: &HashMap<String, Value>,
    assistant_bubble_id: &str,
    assistant_bubble: &HashMap<String, Value>,
) -> Result<()> {
    let tx = conn.transaction()?;
    
    // Ensure chat exists
    ensure_chat_exists_tx(&tx, chat_id)?;
    
    // Save user message
    save_message_to_db_tx(&tx, chat_id, user_bubble_id, user_bubble)?;
    
    // Save assistant message
    save_message_to_db_tx(&tx, chat_id, assistant_bubble_id, assistant_bubble)?;
    
    // Update chat metadata (last updated time)
    let now_ms = chrono::Utc::now().timestamp_millis();
    let mut updates = HashMap::new();
    updates.insert("lastUpdatedAt".to_string(), Value::Number(now_ms.into()));
    update_chat_metadata_tx(&tx, chat_id, updates)?;
    
    tx.commit()?;
    Ok(())
}

